File: services/client/src/protocol/bet_codec.go
package protocol

import (
	"fmt"
	"strconv" // https://pkg.go.dev/strconv
	"strings" // https://pkg.go.dev/strings

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/model"
)

// Utilizo este separador para separar los campos de una bet dentro del payload del mensaje
const (
	fieldSeparator = "|" 
	betSeparator   = "\n"
)

/*
 * Serializa una bet utilizando el formato esperado
 */
func EncodeBet(bet model.Bet) ([]byte, error) {
	if hasReservedSeparator(bet.AgencyId) || hasReservedSeparator(bet.FirstName) || hasReservedSeparator(bet.LastName) || hasReservedSeparator(bet.Birthdate) {
		return nil, fmt.Errorf("bet contains a reserved separator")
	}

	// Convierto todos los campos a string
	fields := []string{
		bet.AgencyId,
		bet.FirstName,
		bet.LastName,
		strconv.FormatUint(bet.Id, 10),
		bet.Birthdate,
		strconv.FormatUint(uint64(bet.Number), 10),
	}

	// Uno todos los campos en un solo string, separados por el separador 
	payload := []byte(strings.Join(fields, fieldSeparator))

	if len(payload) > MaxPayloadSize {
		return nil, fmt.Errorf("bet exceeds maximum payload size")
	}

	return payload, nil
}

/*
 * Deserializa las bets recibidas dentro de un mensaje de resultados
 *
 * Un payload vacio representa que no hubo ganadores
 */
func DecodeBets(payload []byte) ([]model.Bet, error) {
	// No hubo ganadores, el payload es vacio
	if len(payload) == 0 {
		return nil, nil
	}

	// Pueden haber varias bets ganadoras
	encodedBets := strings.Split(string(payload), betSeparator)
	// Creo un slice para almacenar las bets (bets, vacio, largo de la cantidad de apuestas)
	bets := make([]model.Bet, 0, len(encodedBets))

	// Deserializo cada apuesta individual
	for _, encodedBet := range encodedBets {
		// Decodifico cada bet
		bet, err := decodeBet(encodedBet)
		if err != nil {
			return nil, err
		}

		bets = append(bets, bet)
	}

	return bets, nil
}

/*
 * Deserializa una bet individual.
 */
func decodeBet(data string) (model.Bet, error) {
	// Separo los campos usando el separador
	fields := strings.Split(data, fieldSeparator)

	if len(fields) != 6 {
		return model.Bet{}, fmt.Errorf(
			"invalid bet: expected 6 fields, received %d",
			len(fields),
		)
	}

	// Campo, base 10, y 64 bits
	id, err := strconv.ParseUint(fields[3], 10, 64)
	if err != nil {
		return model.Bet{}, fmt.Errorf("invalid Id: %w", err)
	}

	number, err := strconv.ParseUint(fields[5], 10, 32)
	if err != nil {
		return model.Bet{}, fmt.Errorf("invalid bet number: %w", err)
	}

	return model.Bet{
		AgencyId:  fields[0],
		FirstName: fields[1],
		LastName:  fields[2],
		Id:  id,
		Birthdate: fields[4],
		Number:    uint32(number),
	}, nil
}

/*
 * Verifica que un campo no utilice los separadores
 */
func hasReservedSeparator(value string) bool {
	return strings.Contains(value, fieldSeparator) ||
		strings.Contains(value, betSeparator)
}

File: services/server/src/protocol/bet_codec.py
from lottery import Bet


FIELD_SEPARATOR = "|"
BET_SEPARATOR = b"\n"
AMOUNT_OF_FIELDS_IN_BET = 6


"""
Deserializa una bet dando el formato esperado por Lottery

Ej: Recibo "0|Santiago Lionel|Lorca|30904465|1999-03-17|7574"
"""
def decode_bet(payload: bytes) -> Bet:
    try:
        fields = payload.decode("utf-8").split(FIELD_SEPARATOR)
    except UnicodeDecodeError as error:
        raise ValueError("invalid bet encoding")

    # Si no cumpe con el largo de campos esperados retorno error
    if len(fields) != AMOUNT_OF_FIELDS_IN_BET:
        raise ValueError(f"invalid bet: expected {AMOUNT_OF_FIELDS_IN_BET} fields, " f"received {len(fields)}")

    try:
        agency_id = int(fields[0])
        document = int(fields[3])
        number = int(fields[5])
    except ValueError as error:
        raise ValueError("invalid numeric bet field")

    return Bet(
        agency_id = agency_id,
        first_name = fields[1],
        last_name = fields[2],
        document = document,
        birthdate = fields[4],
        number = number,
    )


"""
Deserializa un batch de bets

Cada bet se encuentra separada mediante un salto de linea.
"""
def decode_bets(payload: bytes) -> list[Bet]:
    if not payload:
        raise ValueError("empty bet batch")

    # Separo las distintas bets
    encoded_bets = payload.split(BET_SEPARATOR)

    bets = []

    for encoded_bet in encoded_bets:
        bet = decode_bet(encoded_bet)
        bets.append(bet)

    return bets


"""
Serializa una bet individual
"""
def encode_bet(bet: Bet) -> bytes:
    fields = [
        str(bet.agency_id),
        bet.first_name,
        bet.last_name,
        str(bet.document),
        bet.birthdate,
        str(bet.number),
    ]

    return FIELD_SEPARATOR.join(fields).encode("utf-8")


"""
Serializa las bets ganadoras dentro del payload de RESULTS.

Cada bet se separa mediante un salto de linea
"""
def encode_bets(bets: list[Bet]) -> bytes:
    return BET_SEPARATOR.join(
        encode_bet(bet) for bet in bets
    )